### PR TITLE
Signal subgraph

### LINKED ISSUE #120: signal subgraph
there are now 3 papers that due "signal subgraph"

- https://arxiv.org/abs/1804.09567
- https://arxiv.org/abs/1108.1427
- https://arxiv.org/abs/1801.07683v1

let's have at least some way of doing it here.

### PR DESCRIPTION
**Pull request by vikramc1**
#### Reference Issues/PRs

Fixes #120 

#### What does this implement/fix? Explain your changes.

This implements iterative and one-time vertex screening for finding the signal subgraph with both MGC and DCorr as distance metrics. Added a BaseSignalSubgraph Class in addition to a VertexScreener class. This should be compliant with the sklearn API.

Also added functions to estimate sbm and dcsbm parameters but those are functions and not implemented as classes.

### CODE REVIEW

**Comment by j1c** [comment]
@vikramc1 I would remove the SBM estimator stuff from this PR. @bdpedigo is working on a more coherent submodule for fitting data to different random graph models.

Ill take a more closer look later.

File: graspy/classify/base.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from sklearn.utils.validation import check_is_fitted
+
+
+class BaseSignalSubgraph(ABC, BaseEstimator, ClassifierMixin):
+    """
+    Base Signal Subgraph class.
+    """
+
+    def __init__(self,):
+        self.n_components = n_components
```
Comment (j1c) [inline graspy/classify/base.py]: These lines are uncessary

File: graspy/classify/vertex_screen.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            # ths.append(t_k)
+            k += 1
+        ts = []
+        for i in range(len(V)):
+            A_vk = np.reshape(A[:, V[i][:, None], V[i]], (A.shape[0], -1))
+            ts.append(corr(A_vk, Y)[0])
+        return V, ts
+
+    def _fit_transform(self, graph, y):
+        return self.vertices_of_interest
```
Comment (j1c) [inline graspy/classify/vertex_screen.py]: This doesn't actually fit prior to doing its thing.

File: graspy/classify/vertex_screen.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
+from .base import BaseSignalSubgraph
+from ..utils import import_graph, is_symmetric
+from mgcpy.independence_tests.dcorr import DCorr
+from mgcpy.independence_tests.mgc import MGC
+import numpy as np
+
+
+class VertexScreener(BaseSignalSubgraph):
```
Comment (j1c) [inline graspy/classify/vertex_screen.py]: Docstrings please.

File: graspy/classify/__init__.py [hunk 4]
```diff
@@ -0,0 +1 @@
+__all__ = []
```
Comment (j1c) [inline graspy/classify/__init__.py]: Add "VertexScreener" in square brackets (as string)

File: graspy/classify/vertex_screen.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    def __init__(self, num_vertices=20, distance_metric="mgc"):
+        self.distance_metric = distance_metric
+        self.num_vertices = num_vertices
+
+    def fit(self, graphs, y):
+        self._vertices, self._ts = self._screen_vertices(graphs, y)
```
Comment (j1c) [inline graspy/classify/vertex_screen.py]: Some type checks? E.g. make sure size of graphs and y match, graph is actually a graph (e.g. square, numeric, etc) - you can use import_graph here, I believe `mgc` requires `y` to be a 2d array.. Im sure I missed other checks, but think about other checks here.

File: graspy/classify/vertex_screen.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        V = [np.arange(len(A[-1]))]
+        # while the vertex list is not empty
+        while len(V[k]) > 1:
+            # get the correlation between the current vertex
+            # list and the labels
+            c = self._get_distance_correlations(A[:, V[k][:, None], V[k]], Y)
+            # t_k = np.quantile(c,delta)
+            idx = np.argmin(c)
+            V.append(V[k][np.arange(len(V[k])) != idx])
+            # if none of the vertices are above the threshold
```
Comment (j1c) [inline graspy/classify/vertex_screen.py]: Remove unnecessary comments.

File: graspy/classify/vertex_screen.py (hunk 3 above)
Comment (j1c) [reply graspy/classify/vertex_screen.py]: For class and all functions

File: graspy/classify/base.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+        Parameters
+        ----------
+        graph: np.ndarray or networkx.Graph
+
+        y : Ignored
+
+        Returns
+        -------
+        out : np.ndarray, shape (n_vertices, n_dimension) OR tuple (len 2)
```
Comment (j1c) [inline graspy/classify/base.py]: Update docstrings

File: graspy/classify/base.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
+# base.py
+# Created by Vikram Chandrashekhar on 2019-03-05.
+# Email: vikramc@jhmi.edu
+
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+from sklearn.base import BaseEstimator, ClassifierMixin
```
Comment (j1c) [inline graspy/classify/base.py]: Remove imports you don't need

File: graspy/classify/vertex_screen.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+class VertexScreener(BaseSignalSubgraph):
+    def __init__(self, num_vertices=20, distance_metric="mgc"):
+        self.distance_metric = distance_metric
+        self.num_vertices = num_vertices
+
+    def fit(self, graphs, y):
+        self._vertices, self._ts = self._screen_vertices(graphs, y)
+        num_vertices_ = np.array([len(i) for i in self._vertices], dtype="int")
+        self.vertices_of_interest = self._vertices[
```
Comment (j1c) [inline graspy/classify/vertex_screen.py]: `self.vertices_of_interest` -> `self.vertices_of_interest_`

**Comment by j1c** [comment]
I guess this only computes the subgraph given population of input graphs so Im thinking whether this actually belongs in a `classify` submodule. There is no actual classification happening here.

Let me think

**Comment by bdpedigo** [comment]
More of an estimator?

**Comment by j1c** [comment]
Yea I guess? Or maybe a transformation?

File: graspy/classify/base.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
+# base.py
```
Comment (bdpedigo) [inline graspy/classify/base.py]: make header compliant with the new headers (can just copy and paste from a file in master)

File: graspy/classify/base.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        Returns
+        -------
+        self : BaseSignalSubgraph object
+            Contains the signal subgraph vertices.
+
+        See Also
+        --------
+        
+        """
+        # call self._reduce_dim(A) from your respective embedding technique.
```
Comment (bdpedigo) [inline graspy/classify/base.py]: remove

**Pull request closed without merging**